### ISSUE TITLE
Problem during concurrent tests

### ISSUE THREAD

**Issue by giovannipizzi** [giovannipizzi/disk-objectstore#4]
I got at least in one case (ubuntu, py3.5) an error on the concurrent tests.
The error was from one of the workers, that raised this error: 
https://github.com/giovannipizzi/disk-objectstore/blob/af8e70f2c168f4df3ec3fdf70ffeb9609577bf8c/disk_objectstore/utils.py#L304-L307

If this appears again, we need to better benchmark it (possibly printing the output of *all* concurrent jobs, so we can check what happened at the same time.

**Comment by giovannipizzi** [comment]
Note that in f80efcec56371d898f79ff335390359a14ace3b8 I slightly changed the logic after I saw the error - I am not sure, though, that this should change the behaviour

**Comment by giovannipizzi** [comment]
See also messages in #25

**Comment by giovannipizzi** [comment]
Raised as 'additional features' because for now we can indicate that concurrent packing is not supported, but I would like to check if we can lift this requirement.

**Comment by giovannipizzi** [comment]
In a recent test (with a slightly changed output), I get
```
AssertionError: Mismatch for d7f69ad18a481e84250ef61e8bcbd9817914b98f45bed244b2e68f1f7a9c95ce: d7f69ad18a481e84250ef61e8bcbd9817914b98f45bed244b2e68f1f7a9c95ce vs. e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Note that `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855` is the sha256 of an empty string.
And if I don't delete the repository, and check the object with hash key `d7f69ad18a481e84250ef61e8bcbd9817914b98f45bed244b2e68f1f7a9c95ce`, it's there and it's not corrupted.

So I think what is happening is that the library returns an empty string when it shouldn't. Possible cause: it starts reading a loose object, and while it's being read, it also gets packed and deleted, and it returns an empty string? To be further investigated.

**Comment by giovannipizzi** [comment]
The problem was not what I describe above. It was that the SQLite was committed *before* closing and flushing the pack file.
I addressed this in 4d062ba4bc79c4fd9bed4fdbdb3911f410ec1054.

We need to check if this now works correctly.

IMPORTANT NOTE:
- On POSIX, if I delete a file while it's open, the process can still read it.
- On Windows, this is not the case. Check what happens in the periodic packer if the loose object is deleted after packing while being read.

So, for windows, it's now the packer that might fail!
A good solution, if this happens, is just to ignore the exception and leave the loose object where it is. This is valid; the object will be read from the packs from now on, and the next packing will delete it.

**Comment by giovannipizzi** [comment]
We might want to make the concurrent tests run for longer to spot any other issue

**Comment by giovannipizzi** [comment]
Note that the very first problem was due to the same issue described earlier, i.e., getting zero bytes, because we were accessing the pack before it was flushed, in the case the object was stored as compressed.

This can be reproduced as:
`disk_objectstore.utils.StreamDecompresser(io.BytesIO()).read()`
that returns
`ValueError: There is no data in the reading buffer, but we didn't reach the end of the compressed stream: there must be a problem in the incoming buffer`.
Therefore, I consider also that as solved.


Regarding instead the note in my earlier comment:
> So, for windows, it's now the packer that might fail!
> A good solution, if this happens, is just to ignore the exception and leave the loose object where it is. This is valid; the object will be read from the packs from now on, and the next packing will delete it.

This is fixed in 07382cd92040d0033bee537a8de55fa78d7830ff

However, I got randomly this error in one of the runners:
```
** STDERR OF WORKER 3
Traceback (most recent call last):
  File "D:\a\disk-objectstore\disk-objectstore\tests\concurrent_tests\periodic_worker.py", line 192, in <module>
    main()  # pylint: disable=no-value-for-parameter
  File "c:\hostedtoolcache\windows\python\3.5.4\x64\lib\site-packages\click\core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "c:\hostedtoolcache\windows\python\3.5.4\x64\lib\site-packages\click\core.py", line 782, in main
    rv = self.invoke(ctx)
  File "c:\hostedtoolcache\windows\python\3.5.4\x64\lib\site-packages\click\core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\hostedtoolcache\windows\python\3.5.4\x64\lib\site-packages\click\core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "D:\a\disk-objectstore\disk-objectstore\tests\concurrent_tests\periodic_worker.py", line 137, in main
    retrieved_content = container.get_objects_content(all_hashkeys, skip_if_missing=False)
  File "d:\a\disk-objectstore\disk-objectstore\disk_objectstore\container.py", line 552, in get_objects_content
    for obj_hashkey, stream, _ in triplets:
  File "d:\a\disk-objectstore\disk-objectstore\disk_objectstore\container.py", line 465, in get_object_stream_generator
    last_open_file = open(obj_path, mode='rb')
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpb542gz99\\container\\loose\\2c\\4aa1cf7438e6fddb5c6c6996a4fb89f01702f6ae9b6ce8314dc14c50545eab'
```
that needs to be investigated. This is in the worker, while trying to read files.
The object should have been already created and synced. So this is probably happening while the file is being deleted, maybe at the same time? It is strange that this gives a PermissionError however, maybe it happens for concurrent deletion and opening?
